### PR TITLE
Fixed #32062 -- Added %b support to Date.strftime.

### PR DESCRIPTION
**Pull request by pandafy**
Added "%b" option in "Date.strftime" method to render abbreviated
month name in Associated Press style

### CODE REVIEW

**Comment by adamchainz** [comment]
Why did you use `%b` for this AP style? It doesn't seem consistent with Python or Django.

`%b` in Python's datetime is "Month as locale’s abbreviated name." which makes returns "Mar" for March. Django's `date` template filter [adds "N" as the Associated Press style](https://docs.djangoproject.com/en/3.0/ref/templates/builtins/#std:templatefilter-date) which returns "March", as you have here in the code.

**Comment by pandafy** [comment]
Thanks for pointing that out Adam. I overlooked the difference between `%b` and `%N`. I went with adding AP style because I thought it would be more consistent with what Django admin already displays for dates and it will be better to be able to format dates like that. 

`%b` in Django's `date` template formats month in 3 characters lowercase format, e.g. `jan, feb`
`%b` in Python's `datetime`  formats month as abbreviated month name by capitalising first character, e.g. `Jan, Feb`

I think it will be better to follow the Python's conventions of `datetime` implementation since `%B` option already does that.

Should I also add `%N` option as it can come handy in future?

File: django/contrib/admin/static/admin/js/core.js [hunk 1]
```diff
@@ -85,6 +85,18 @@ function findPosY(obj) {
         return (this.getSeconds() < 10) ? '0' + this.getSeconds() : this.getSeconds();
     };
 
+    Date.prototype.getAbbrevMonthName = function() {
+        return typeof window.CalendarNamespace === "undefined"
+            ? this.getTwoDigitMonth()
+            : this.getFullMonthName().substring(0, 3);
```
Comment (adamchainz) [inline django/contrib/admin/static/admin/js/core.js:91]: I don't think this will work in all languages. Probably there should be two sets of translations.

**Review by adamchainz — COMMENTED** [review]
I think it's best to stick to adding only the abbreviated style in this PR.

What about documentation for this change, is there not a place in the documentation this is represented?

**Comment by felixxm** [comment]
@pandafy Thanks for this patch 👍 Please limit this PR only to the "%b" format, you shouldn't add features not accepted in the ticket.

**Comment by pandafy** [comment]
> @pandafy Thanks for this patch  Please limit this PR only to the "%b" format, you shouldn't add features not accepted in the ticket.

I have removed the implementation for "%N" option. I will always keep this in mind to limit the scope of a PR to only thing accepted in the ticket while making contributions in future. :smile: 

> What about documentation for this change, is there not a place in the documentation this is represented?

I failed to find documentation for implementation of `strftime` in JavaScript. It will be really helpful if you can point me to the documentation. I searched for `strftime` in the `docs` directory and looked at commits adding `strftime` in code. 

I think it was supposed to be just used internally in Django and not to be offered to users. Since, it can be useful to users(at least it was to me) it will be worth documenting. Should I create a new ticket for documenting `strftime` and `strptime` ?

File: django/contrib/admin/static/admin/js/core.js [hunk 2]
```diff
@@ -85,6 +85,12 @@ function findPosY(obj) {
...
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/core.js:91]:
```suggestion
            : window.CalendarNamespace.monthOfYearAbbrev[this.getMonth()];
```

File: django/contrib/admin/static/admin/js/calendar.js [hunk 3]
```diff
@@ -21,6 +21,20 @@ depends on core.js for utility functions like removeChildren or quickElement
             gettext('November'),
             gettext('December')
         ],
+        abbrevMonthOfYear: [
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/calendar.js:24]:
```suggestion
        monthOfYearAbbrev: [
```

File: django/contrib/admin/static/admin/js/calendar.js [hunk 4]
```diff
@@ -21,6 +21,20 @@ depends on core.js for utility functions like removeChildren or quickElement
...
+            pgettext('abbrev. month January', 'Jan'),
+            pgettext('abbrev. month February', 'Feb'),
+            pgettext('abbrev. month March', 'Mar'),
+            pgettext('abbrev. month April', 'Apr'),
+            pgettext('abbrev. month May', 'May'),
+            pgettext('abbrev. month June', 'Jun'),
+            pgettext('abbrev. month June', 'Jul'),
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/calendar.js:31]:
```suggestion
            pgettext('abbrev. month July', 'Jul'),
```

File: js_tests/admin/core.test.js [hunk 5]
```diff
@@ -40,10 +40,21 @@ QUnit.test('Date.getTwoDigitSecond', function(assert) {
...
+QUnit.test('Date.getAbbrevMonthName', function(assert) {
+    assert.equal(new Date(2020, 0, 26).getAbbrevMonthName(), 'Jan', 'jan 26');
+    assert.equal(new Date(2020, 9, 26).getAbbrevMonthName(), 'Oct', 'oct 26');
+});
+
+QUnit.test('Date.getFullMonthName', function(assert) {
+    assert.equal(new Date(2020, 0, 26).getFullMonthName(), 'January', 'jan 26');
+    assert.equal(new Date(2020, 9, 26).getFullMonthName(), 'October', 'oct 26');
+});
+
```
Comment (felixxm) [inline js_tests/admin/core.test.js:52]: These tests are not necessary, IMO.

**Review by felixxm — COMMENTED** [review]
@pandafy Thanks :+1: 

> What about documentation for this change, is there not a place in the documentation this is represented?

No, docs changes are not required.

File: js_tests/admin/core.test.js (hunk 5 above)
Comment (pandafy) [reply js_tests/admin/core.test.js:52]: @felixxm I asked about adding these tests on [django-core-mentorship mailing list](https://groups.google.com/g/django-core-mentorship/c/r4JSPAiFuLU/m/kRTtrOoGAwAJ). @adamchainz suggested to add tests for individual code paths and I agree with Adam on adding individual tests for each individual method.

File: django/contrib/admin/static/admin/js/core.js (hunk 2 above)
Comment (pandafy) [reply django/contrib/admin/static/admin/js/core.js]: I remember I changed this, should have double checked on GitHub :sweat_smile:

File: js_tests/admin/core.test.js (hunk 5 above)
Comment (felixxm) [reply js_tests/admin/core.test.js:52]: We can keep them, I moved unrelated tests to a separate commit.

**Comment by felixxm** [comment]
@pandafy Welcome aboard :boat:

**Comment by felixxm** [comment]
buildbot, test on selenium.

**Comment by adamchainz** [comment]
👍 Thanks @pandafy and @felixxm !

**Pull request merged**